dataframely/schema.py

```diff
@@ -446,71 +450,192 @@ def _sampling_overrides(cls) -> dict[str, pl.Expr]:
 
     # ---------------------------------- VALIDATION ---------------------------------- #
 
+    @overload
     @classmethod
     def validate(
-        cls, df: pl.DataFrame | pl.LazyFrame, /, *, cast: bool = False
-    ) -> DataFrame[Self]:
-        # We can dispatch to the `filter` method and raise an error if any row cannot
-        # be validated
-        df_valid, failures = cls.filter(df, cast=cast)
-        if len(failures) > 0:
-            raise RuleValidationError(failures.counts())
-        return df_valid
+        cls,
+        df: pl.DataFrame | pl.LazyFrame,
+        /,
+        *,
+        cast: bool = False,
+        eager: Literal[True] = True,
+    ) -> DataFrame[Self]: ...
+
+    @overload
+    @classmethod
+    def validate(
+        cls,
+        df: pl.DataFrame | pl.LazyFrame,
+        /,
+        *,
+        cast: bool = False,
+        eager: Literal[False],
+    ) -> LazyFrame[Self]: ...
+
+    @overload
+    @classmethod
+    def validate(
+        cls,
+        df: pl.DataFrame | pl.LazyFrame,
+        /,
+        *,
+        cast: bool = False,
+        eager: bool,
+    ) -> DataFrame[Self] | LazyFrame[Self]: ...
```

Comment:
why do we need this overload given that `eager: bool` always has a default `True` (i.e., cannot be unset)?